### ISSUE TITLE
Inconsistent HttpHeadersNonValidated behavior when response header added to a request

### ISSUE THREAD

**Issue by pakrym** [dotnet/runtime#62024]
### Description

When adding a response header like `WWW-Authenticate` to the request message `HttpHeadersNonValidated` returns it when enumerated but doesn't when accessed using indexer or `Contains`.

### Reproduction Steps

```c#
HttpRequestMessage requestMessage = new HttpRequestMessage();
requestMessage.Headers.TryAddWithoutValidation("WWW-Authenticate", "Basic realm=\"Access to the staging site\", charset=\"UTF-8\"");
Console.WriteLine(requestMessage.Headers.NonValidated.ToArray()[0].Key);
Console.WriteLine(requestMessage.Headers.NonValidated.Contains("WWW-Authenticate"));
```

```
dotnet run
```

### Expected behavior

```
WWW-Authenticate
True
```

### Actual behavior

```
WWW-Authenticate
False
```

### Configuration

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  3.0.103 [C:\Program Files\dotnet\sdk]
  3.1.120 [C:\Program Files\dotnet\sdk]
  3.1.200-preview-014883 [C:\Program Files\dotnet\sdk]
  3.1.415 [C:\Program Files\dotnet\sdk]
  5.0.209 [C:\Program Files\dotnet\sdk]
  5.0.403 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.1.21451.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.1.21451.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by stephentoub** [comment]
The issue is HttpHeadersNonValidated is using the wrong method to get at the HeaderDescriptor. Basically this:
https://github.com/dotnet/runtime/blob/7cfcbb0fe6c786ce9f6ab11e8edfd40e342ea7d0/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeadersNonValidated.cs#L35
(and other TryGet calls in that same file) are calling to this:
https://github.com/dotnet/runtime/blob/7cfcbb0fe6c786ce9f6ab11e8edfd40e342ea7d0/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs#L47
but should be calling to this:
https://github.com/dotnet/runtime/blob/7cfcbb0fe6c786ce9f6ab11e8edfd40e342ea7d0/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1054
on the associated collection. As a result, it ends up looking for the header with a response descriptor when it was actually added with a custom descriptor.

**Comment by stephentoub** [comment]
re-opening to track potential servicing

**Comment by karelz** [comment]
Triage: New API not working correctly. We should fix it in 6.0 servicing.

**Comment by karelz** [comment]
Fixed in main (7.0) in PR #62150 and in 6.0.2 in PR #62194.
